### PR TITLE
Add tautomer query to the substructlibrary

### PR DESCRIPTION
**Pull request by bp-kelley**
TODO:

Still need to build the java wrappers.

Note contains the fixes from #3798

### CODE REVIEW

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 1]
```diff
@@ -122,6 +142,8 @@ void SubSearcher(const ROMol &in_query, const Bits &bits,
     //  returned molecule!
     const boost::shared_ptr<ROMol> &m = mols.getMol(idx);
     ROMol *mol = m.get();
+    if(!mol)
+      continue;
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:146]:
```suggestion
    if(!mol){
      continue;
    }
```

File: Code/GraphMol/SubstructLibrary/PatternFactory.cpp [hunk 2]
```diff
@@ -19,22 +19,37 @@
...
 		  unsigned int start, unsigned int end, unsigned int numThreads) {
   for (unsigned int idx = start;
        idx < end;
        idx += numThreads) {
-    fps[idx] = fph.makeFingerprint( *slib.getMol(idx).get() );
+    auto mol = slib.getMol(idx);
+    if (mol.get()) {
+        fps[idx] = fph.makeFingerprint( *mol.get() );
+    }
+    else {
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/PatternFactory.cpp:32]:
```suggestion
    } else {
```
(general aside: wouldn't be bad to run clang-format pre-submission)

**Review by greglandrum — CHANGES_REQUESTED** [review]
- Unless I missed something, the only tests that are here are for cases where the TautomerQueries behave the same as normal queries. You need some tests for cases that show the tautomer queries are actually working and are returning different results than a standard query.
- The conflicts are because you've incorporated the changes that were in #3798. 
- There are some other minor changes suggested below.

File: Code/GraphMol/SubstructLibrary/PatternFactory.cpp (hunk 2 above)
Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/PatternFactory.cpp:32]: Now that #3798 is accepted I’ll merge with master.

I’ll add the extra tests, I misunderstood your previous comment.

**Comment by bp-kelley** [comment]
@greglandrum  I'm ready to go back to this one :). I'll let you know when I have fixed the conflicts.

**Comment by bp-kelley** [comment]
@greglandrum two really stupid bugs.  Always look at the warnings :)

File: Code/GraphMol/SubstructLibrary/PatternFactory.h [hunk 3]
```diff
@@ -19,7 +19,13 @@ namespace RDKit {
 /*
      \param sslib The substructure library (without pattern fingerprints)
      \param numThreads the number of threads to use, -1 for all threads [default 1]
+     \param patterns if supplied, use this pattern holder and take ownership [default nullptr]
 */
-RDKIT_SUBSTRUCTLIBRARY_EXPORT void addPatterns(SubstructLibrary &sslib, int numThreads=1);  
-}
+RDKIT_SUBSTRUCTLIBRARY_EXPORT void addPatterns(SubstructLibrary &sslib,
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/PatternFactory.h:24]: Please swap the order of these two declarations. The current ordering ends up making the documentation wrong.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 4]
```diff
@@ -69,6 +67,22 @@ struct Bits {
...
+  Bits(const FPHolderBase *fps, const TautomerQuery &m, bool recursionPossible,
+       bool useChirality, bool useQueryQueryMatches)
+      : fps(fps),
+        recursionPossible(recursionPossible),
+        useChirality(useChirality),
+        useQueryQueryMatches(useQueryQueryMatches) {
+    if (fps) {
+      const TautomerPatternHolder *tp =
+          dynamic_cast<const TautomerPatternHolder *>(fps);
+      PRECONDITION(tp, "TautomerQueries only support TautomerPatternHolders");
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:79]: Not sure if this should be a PRECONDITION (equivalent to an assertion; something which indicates a serious programming error) or maybe just throw a `ValueErrorException` (it's a mistake, but one which can be recovered from)

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.h [hunk 5]
```diff
@@ -461,118 +482,187 @@ class RDKIT_SUBSTRUCTLIBRARY_EXPORT SubstructLibrary {
...
+  //! Return the number of matches for the query
+  /*!
+    \param query       TautomereQuery to match against molecules
+    \param recursionPossible  flags whether or not recursive matches are allowed [ default true ]
+    \param useChirality  use atomic CIP codes as part of the comparison [ default true ]
+    \param useQueryQueryMatches  if set, the contents of atom and bond queries 
+                                 will be used as part of the matching [ default false ]
+    \param numThreads  If -1 use all available processors [default -1]
+  */
+  unsigned int countMatches(const TautomerQuery &query,
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.h:570]: Couldn't you just have this and the overload above be a single template method?
```
template <typename QueryType>
  unsigned int countMatches(const QueryType &query, bool recursionPossible = true,
                            bool useChirality = true,
                            bool useQueryQueryMatches = false,
                            int numThreads = -1) const {
        return countMatches(query,0,mols->size(),recursionPossible,useChirality,useQueryQueryMatches,numThreads);

    }
```

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 6]
```diff
@@ -304,6 +344,15 @@ unsigned int SubstructLibrary::countMatches(const ROMol &query,
                       useQueryQueryMatches, numThreads);
 }
 
+unsigned int SubstructLibrary::countMatches(const TautomerQuery &query,
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:347]: See my comment in the header about making this (and the overload above it) a simple template which is defined in the header.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 7]
```diff
@@ -322,6 +380,16 @@ bool SubstructLibrary::hasMatch(const ROMol &query, bool recursionPossible,
              .size() > 0;
 }
 
+bool SubstructLibrary::hasMatch(const TautomerQuery &query,
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:383]: This (and the overload above it) could also be template methods defined in the header.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 8]
```diff
@@ -332,13 +400,24 @@ bool SubstructLibrary::hasMatch(const ROMol &query, unsigned int startIdx,
...
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:403]: same comment about making this a template

File: Code/GraphMol/TautomerQuery/TautomerQuery.h [hunk 9]
```diff
@@ -74,7 +74,7 @@ class RDKIT_TAUTOMERQUERY_EXPORT TautomerQuery {
 
   // Query fingerprint
   ExplicitBitVect *patternFingerprintTemplate(unsigned int fpSize = 2048U);
```
Comment (greglandrum) [inline Code/GraphMol/TautomerQuery/TautomerQuery.h:76]: Do you need the non-`const` version of this method at all?
The two implementations look identical.

Comment (greglandrum) [reply Code/GraphMol/TautomerQuery/TautomerQuery.h:76]: might as well just remove the non-const one.

**Review by greglandrum — CHANGES_REQUESTED** [review]
Suggestions are all related to code cleanup.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 10]
```diff
@@ -30,14 +30,13 @@
 //
 #include "SubstructLibrary.h"
 #include <RDGeneral/RDThreads.h>
-#ifdef RDK_THREADSAFE_SSS
+//#ifdef RDK_THREADSAFE_SSS
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:33]: nope, good catch.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp (hunk 4 above)
Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:79]: I just made it a warning.

File: Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp [hunk 11]
```diff
@@ -90,15 +104,15 @@ unsigned int SubstructLibrary::addMol(const ROMol &m) {
 namespace {
 //  Return true if the pattern contains a ring query
 bool query_needs_rings(const ROMol &in_query) {
-  for (auto &atom: in_query.atoms()) {
-    if(atom->hasQuery()) {
+  for (auto &atom : in_query.atoms()) {
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp:107]: I think auto here has to be const actually but I'll make it more specific.

File: Code/GraphMol/TautomerQuery/TautomerQuery.h (hunk 9 above)
Comment (bp-kelley) [reply Code/GraphMol/TautomerQuery/TautomerQuery.h:76]: We need the non-const one for serialization.

**Comment by bp-kelley** [comment]
WIth the template change I'll need to fix the java api.

**Comment by bp-kelley** [comment]
@greglandrum java should be fixed now

**Review by greglandrum — APPROVED** [review]

**Pull request merged**